numpy/core/include/numpy/pythoncapi_compat.h

```diff
@@ -0,0 +1,342 @@
+// Header file providing new functions of the Python C API to old Python
+// versions.
+//
+// File distributed under the MIT license.
+//
+// Homepage:
+// https://github.com/pythoncapi/pythoncapi_compat
+//
+// Latest version:
+// https://raw.githubusercontent.com/pythoncapi/pythoncapi_compat/master/pythoncapi_compat.h
+//
+// SPDX-License-Identifier: MIT
+
+#ifndef PYTHONCAPI_COMPAT
+#define PYTHONCAPI_COMPAT
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <Python.h>
+#include "frameobject.h"          // PyFrameObject, PyFrame_GetBack()
+
+
+// Compatibility with Visual Studio 2013 and older which don't support
+// the inline keyword in C (only in C++): use __inline instead.
+#if (defined(_MSC_VER) && _MSC_VER < 1900 \
+     && !defined(__cplusplus) && !defined(inline))
+#  define inline __inline
+#  define PYTHONCAPI_COMPAT_MSC_INLINE
+   // These two macros are undefined at the end of this file
+#endif
+
+
+// Cast argument to PyObject* type.
+#ifndef _PyObject_CAST
+#  define _PyObject_CAST(op) ((PyObject*)(op))
+#endif
+#ifndef _PyObject_CAST_CONST
+#  define _PyObject_CAST_CONST(op) ((const PyObject*)(op))
+#endif
+
+
+// bpo-42262 added Py_NewRef() to Python 3.10.0a3
+#if PY_VERSION_HEX < 0x030A00A3 && !defined(Py_NewRef)
+static inline PyObject* _Py_NewRef(PyObject *obj)
+{
+    Py_INCREF(obj);
+    return obj;
+}
+#define Py_NewRef(obj) _Py_NewRef(_PyObject_CAST(obj))
+#endif
+
+
+// bpo-42262 added Py_XNewRef() to Python 3.10.0a3
+#if PY_VERSION_HEX < 0x030A00A3 && !defined(Py_XNewRef)
+static inline PyObject* _Py_XNewRef(PyObject *obj)
+{
+    Py_XINCREF(obj);
+    return obj;
+}
+#define Py_XNewRef(obj) _Py_XNewRef(_PyObject_CAST(obj))
+#endif
+
+
+// See https://bugs.python.org/issue42522
+#if !defined(_Py_StealRef)
+static inline PyObject* __Py_StealRef(PyObject *obj)
+{
+    Py_DECREF(obj);
+    return obj;
+}
+#define _Py_StealRef(obj) __Py_StealRef(_PyObject_CAST(obj))
+#endif
+
+
+// See https://bugs.python.org/issue42522
+#if !defined(_Py_XStealRef)
+static inline PyObject* __Py_XStealRef(PyObject *obj)
+{
+    Py_XDECREF(obj);
+    return obj;
+}
+#define _Py_XStealRef(obj) __Py_XStealRef(_PyObject_CAST(obj))
+#endif
+
+
+// bpo-39573 added Py_SET_REFCNT() to Python 3.9.0a4
+#if PY_VERSION_HEX < 0x030900A4 && !defined(Py_SET_REFCNT)
+static inline void _Py_SET_REFCNT(PyObject *ob, Py_ssize_t refcnt)
+{
+    ob->ob_refcnt = refcnt;
+}
+#define Py_SET_REFCNT(ob, refcnt) _Py_SET_REFCNT(_PyObject_CAST(ob), refcnt)
+#endif
```

Comment:
Why is this a thing? It will horribly break non-CPython implementations.

Reply:
@mattip I see that it is already used in `numpy/core/src/multiarray/array_coercion.c`, is that in error?

Reply:
Either:
- all the fields are guaranteed to be zeroed out in that allocation, in which case we could use `Py_INCREF` instead, or
- we could be returning random data in other fields of the struct.

So I think use of that "convenience function" could enable masking possible bugs.

Reply:
> Why is this a thing? It will horribly break non-CPython implementations.

The _PyObject_CAST() is required to prevent introducing new compiler warnings when ``Py_REFCNT(obj) = refcnt;`` is replaced with ``Py_SET_REFCNT(obj, refcnt);``. Many functions of the Python C API were implemented as macro which casted their argument to ``PyObject*``, but static inline functions define a type for their parameters (usually ``PyObject*``).

In Python 3.10, ``Py_REFCNT(obj) = refcnt;`` now fails with a compiler error: Py_SET_REFCNT() must be used. See:

* https://docs.python.org/dev/c-api/structures.html#c.Py_REFCNT
* https://vstinner.github.io/c-api-opaque-structures.html

Reply:
I didn't realize it was added to `npy_3kcompat.h` in #16899.

Reply:
... which was needed for python3.10.

The root cause was 1035c3f830 in #16200 which changed
```
PyArrayObject_fields arr_fields = {
            .ob_base.ob_refcnt = 1,
            .ob_base.ob_type = &PyArray_Type,
             .flags = NPY_ARRAY_WRITEABLE,  /* assume array is not behaved. */
         };
```

to 
```
PyArrayObject_fields arr_fields = {
             .flags = NPY_ARRAY_WRITEABLE,  /* assume array is not behaved. */
         };
    Py_SET_TYPE(&arr_fields, &PyArray_Type);
    Py_REFCNT(&arr_fields) = 1;
```

and `Py_REFCNT` went away in python3.10

Reply:
Well, that spot should really be phased out soon, hopefully (there might be a similar one in other "casting" code that also should be used less soon hopefully).
At that point this would only be necessary as a backward compatibility hack.  And that means that the performance gain of using a stack allocation stops mattering (I am not sure how big it is, but I assume it is pretty significant here).

Reply:
+1 for getting rid of this code.